shared_model/backend/protobuf/impl/proto_query_response_factory.cpp

```diff
@@ -0,0 +1,230 @@
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::QueryResponse>
+shared_model::proto::ProtoQueryResponseFactory::createSignatoriesResponse(
+    std::vector<shared_model::interface::types::PubkeyType> signatories,
+    const crypto::Hash &query_hash) {
+  iroha::protocol::QueryResponse protocol_query_response;
+  protocol_query_response.set_query_hash(query_hash.hex());
+
+  iroha::protocol::SignatoriesResponse *protocol_specific_response =
+      protocol_query_response.mutable_signatories_response();
+  for (const auto &key : signatories) {
+    const auto &blob = key.blob();
+    protocol_specific_response->add_keys(blob.data(), blob.size());
+  }
+
+  return std::make_unique<shared_model::proto::QueryResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::QueryResponse>
+shared_model::proto::ProtoQueryResponseFactory::createTransactionsResponse(
+    std::vector<std::shared_ptr<shared_model::interface::Transaction>>
+        transactions,
+    const crypto::Hash &query_hash) {
+  iroha::protocol::QueryResponse protocol_query_response;
+  protocol_query_response.set_query_hash(query_hash.hex());
+
+  iroha::protocol::TransactionsResponse *protocol_specific_response =
+      protocol_query_response.mutable_transactions_response();
+  for (const auto &tx : transactions) {
+    protocol_specific_response->add_transactions()->CopyFrom(
+        std::static_pointer_cast<shared_model::proto::Transaction>(tx)
+            ->getTransport());
+  }
+
+  return std::make_unique<shared_model::proto::QueryResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::QueryResponse>
+shared_model::proto::ProtoQueryResponseFactory::createAssetResponse(
+    std::unique_ptr<shared_model::interface::Asset> asset,
+    const crypto::Hash &query_hash) {
+  iroha::protocol::QueryResponse protocol_query_response;
+  protocol_query_response.set_query_hash(query_hash.hex());
+
+  iroha::protocol::AssetResponse *protocol_specific_response =
+      protocol_query_response.mutable_asset_response();
+  protocol_specific_response->mutable_asset()->CopyFrom(
+      static_cast<shared_model::proto::Asset *>(asset.get())->getTransport());
+
+  return std::make_unique<shared_model::proto::QueryResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::QueryResponse>
+shared_model::proto::ProtoQueryResponseFactory::createRolesResponse(
+    std::vector<shared_model::interface::types::RoleIdType> roles,
+    const crypto::Hash &query_hash) {
+  iroha::protocol::QueryResponse protocol_query_response;
+  protocol_query_response.set_query_hash(query_hash.hex());
+
+  iroha::protocol::RolesResponse *protocol_specific_response =
+      protocol_query_response.mutable_roles_response();
+  for (const auto &role : roles) {
+    protocol_specific_response->add_roles(role);
```

Comment:
It could be `add_roles(std::move(role))` if `const` is removed above.

Reply:
it also possible to use `auto &&role`, btw